spidermonkey_91 can't be built on i686

### Steps To Reproduce
Steps to reproduce the behavior:
1. ``` nix-build '<nixpkgs>' -A pkgsi686Linux.spidermonkey_91 ```

### Build log
```
In file included from Unified_cpp_js_src_jit9.cpp:11:
/build/firefox-91.10.0/js/src/jit/Recover.cpp:312:11: note: 'result' was declared here
  312 |   int32_t result;
      |           ^~~~~~
make[3]: Leaving directory '/build/firefox-91.10.0/obj/js/src'
make[3]: Leaving directory '/build/firefox-91.10.0/obj/js/src/jit'
    Finished release [optimized] target(s) in 2m 25s
/build/firefox-91.10.0/obj/_virtualenvs/common/bin/python -m mozbuild.action.check_binary --target --networking /build/firefox-91.10.0/obj/i686-unknown-linux-gnu/release/libjsrust.a
BUILDTASK {"argv": ["/build/firefox-91.10.0/python/mozbuild/mozbuild/action/check_binary.py", "--target", "--networking", "/build/firefox-91.10.0/obj/i686-unknown-linux-gnu/release/libjsrust.a"], "start": 1655410272.4902613, "end": 1655410272.5648446, "context": null}
make[3]: Leaving directory '/build/firefox-91.10.0/obj/js/src/rust'
make[2]: Leaving directory '/build/firefox-91.10.0/obj'
make[1]: *** [/build/firefox-91.10.0/config/recurse.mk:34: compile] Error 2
make[1]: Leaving directory '/build/firefox-91.10.0/obj'
make: *** [/build/firefox-91.10.0/config/rules.mk:355: default] Error 2
```


### Notify maintainers
@lostnet 


### Metadata
```console
 - system: `"i686-linux"`
 - host os: `Linux 5.15.16-lqx2, NixOS, 22.11 (Raccoon), 22.11pre385606.01e4d1a67f6`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.9.1`
 - channels(root): `"nixos-stable-22.05"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

@lostnet I've got this this error:

```
In file included from Unified_cpp_js_src_jit2.cpp:20:
/build/firefox-91.10.0/js/src/jit/CacheIR.cpp: In member function 'js::jit::AttachDecision js::jit::GetPropIRGenerator::tryAttachPrimitive(js::jit::ValOperandId, JS::HandleId)':
/build/firefox-91.10.0/js/src/jit/CacheIR.cpp:2128:14: warning: 'protoKey' may be used uninitialized in this function [-Wmaybe-uninitialized]
 2128 |   JSProtoKey protoKey;
      |              ^~~~~~~~
make[3]: Leaving directory '/build/firefox-91.10.0/obj/js/src/wasm'
In file included from Unified_cpp_js_src_jit3.cpp:2:
/build/firefox-91.10.0/js/src/jit/CodeGenerator.cpp: In member function 'void js::jit::CodeGenerator::emitIonToWasmCallBase(js::jit::LIonToWasmCallBase<NumDefs>*) [with unsigned int NumDefs = 1]':
/build/firefox-91.10.0/js/src/jit/CodeGenerator.cpp:15441:26: warning: 'argMir' may be used uninitialized in this function [-Wmaybe-uninitialized]
15441 |     ABIArg arg = abi.next(argMir);
      |                  ~~~~~~~~^~~~~~~~
/build/firefox-91.10.0/js/src/jit/CodeGenerator.cpp: In member function 'void js::jit::CodeGenerator::emitIonToWasmCallBase(js::jit::LIonToWasmCallBase<NumDefs>*) [with unsigned int NumDefs = 2]':
/build/firefox-91.10.0/js/src/jit/CodeGenerator.cpp:15441:26: warning: 'argMir' may be used uninitialized in this function [-Wmaybe-uninitialized]
15441 |     ABIArg arg = abi.next(argMir);
      |                  ~~~~~~~~^~~~~~~~
make[3]: Leaving directory '/build/firefox-91.10.0/obj/js/src/jit'
make[2]: Leaving directory '/build/firefox-91.10.0/obj'
make[1]: *** [/build/firefox-91.10.0/config/recurse.mk:34: compile] Error 2
make[1]: Leaving directory '/build/firefox-91.10.0/obj'
make: *** [/build/firefox-91.10.0/config/rules.mk:355: default] Error 2
```

```console
/nix/store/anc0xd18885xhyp3hrriqn2hq3v3v3ws-gcc-wrapper-10.3.0/bin/g++ -std=gnu++17 -o e_acos.o -c  -I/build/firefox-91.10.0/obj/dist/sy>
modules/fdlibm/src/e_acosh.o
In file included from /build/firefox-91.10.0/modules/fdlibm/src/e_acos.cpp:44:
/build/firefox-91.10.0/modules/fdlibm/src/math_private.h:34:21: error: conflicting declaration 'typedef __double_t double_t'
   34 | typedef __double_t  double_t;
      |                     ^~~~~~~~
In file included from /nix/store/vwggfsx0bjzzdd6bp2phyl342xf01jij-gcc-10.3.0/include/c++/10.3.0/cmath:45,
                 from /build/firefox-91.10.0/obj/dist/system_wrappers/cmath:3,
                 from /build/firefox-91.10.0/modules/fdlibm/src/e_acos.cpp:41:
/nix/store/1xq789b12h7bvc2h64k7vcd71m9jjc7w-glibc-2.34-210-dev/include/math.h:170:21: note: previous declaration as 'typedef long double>
  170 | typedef long double double_t;
      |                     ^~~~~~~~
make[3]: *** [/build/firefox-91.10.0/config/rules.mk:676: e_acos.o] Error 1
```

Upstream issue: https://bugzilla.mozilla.org/show_bug.cgi?id=1729459

--- a/pkgs/development/interpreters/spidermonkey/common.nix
+++ b/pkgs/development/interpreters/spidermonkey/common.nix
@@ -75,6 +75,9 @@ stdenv.mkDerivation (finalAttrs: rec {
       url = "https://hg.mozilla.org/releases/mozilla-esr102/raw-rev/1fa20fb474f5d149cc32d98df169dee5e6e6861b";
       sha256 = "sha256-eCisKjNxy9SLr9KoEE2UB26BflUknnR7PIvnpezsZeA=";
     })
+  ] ++ lib.optionals (lib.versionAtLeast version "91" && stdenv.hostPlatform.system == "i686-linux") [
+    # Fixes i686 build, https://bugzilla.mozilla.org/show_bug.cgi?id=1729459
+    ./fix-float-i686.patch
   ];
 
   nativeBuildInputs = [

--- a/pkgs/development/interpreters/spidermonkey/fix-float-i686.patch
+++ b/pkgs/development/interpreters/spidermonkey/fix-float-i686.patch
@@ -1,0 +1,16 @@
+diff --git a/modules/fdlibm/src/math_private.h b/modules/fdlibm/src/math_private.h
+index 51d79f9c2ec59..fafd7d6fc1e0d 100644
+--- a/modules/fdlibm/src/math_private.h
++++ b/modules/fdlibm/src/math_private.h
+@@ -30,5 +30,9 @@
+  * Adapted from https://github.com/freebsd/freebsd-src/search?q=__double_t
+  */
+
++#if defined __FLT_EVAL_METHOD__ && (__FLT_EVAL_METHOD__ == 2)
++typedef long double      __double_t;
++#else
+ typedef double      __double_t;
++#endif
+ typedef __double_t  double_t;
+
+ /*

